### PR TITLE
fix: fix custom wordbreaker output format for sil.km.gcc

### PR DESCRIPTION
**Pull request by jahorton**
Fixes https://github.com/keymanapp/keyman/issues/12200.

Recent changes in 18.0-alpha require an update to the custom wordbreaker I originally wrote for this model.  We now seek to model whitespaces in the context to ensure we don't accidentally auto-correct away punctuation marks, but those same changes require precise adherence to the wordbreaker output format.  Unfortunately, my original version of it played a bit... fast and loose, returning improper values for blank or duplicated tokens.

### CODE REVIEW

File: release/sil/sil.km.ggc/source/sil.km.ggc.model.ts [hunk 1]
```diff
@@ -10,59 +10,71 @@
...
+      const token = plainTokens[i];
+      if(token.length == 0) {
+        plainTokens.splice(i, 1);
+        i--;
+        continue;
+      } else if(token.length == 1 && whitespaceRegex.test(token)) {
+        plainTokens.splice(i, 1);
+        i--;
+        continue;
+      }
```
Comment (jahorton) [inline release/sil/sil.km.ggc/source/sil.km.ggc.model.ts]: Prevents empty tokens from appearing in the middle of context, especially should excessive whitespaces exist between words.  Something like `the        quick   brown fox` should appear to skip all spaces between tokens at once, not in multiple steps.

File: release/sil/sil.km.ggc/source/sil.km.ggc.model.ts [hunk 2]
```diff
@@ -10,59 +10,71 @@
...
-      // there was a `right` portion, as it may have extra marks that also need to be spun off.
-      i--;
     }
-   }
-   return tokens.map(function(token) {
+
+    let latestIndex = 0;
+    return tokens.map(function(token) {
+      const start = str.indexOf(token, latestIndex);
+      latestIndex = start + token.length;
```
Comment (jahorton) [inline release/sil/sil.km.ggc/source/sil.km.ggc.model.ts:66]: The key part of the fix:  we should track the furthest index reached while iterating over each token and skip past it when determining the original position of the next token to be processed.  This fixes handling for any duplicate-string cases.

File: release/sil/sil.km.ggc/source/sil.km.ggc.model.ts [hunk 3]
```diff
@@ -10,59 +10,71 @@
...
+      const token = tokens[i];
+      if(token.length == 0) {
+        tokens.splice(i, 1);
+        i--;
+        continue;
+      } else if(token.length == 1 && whitespaceRegex.test(token)) {
+        tokens.splice(i, 1);
+        i--;
+        continue;
+      }
```
Comment (jahorton) [inline release/sil/sil.km.ggc/source/sil.km.ggc.model.ts:27]: Prevents empty tokens from appearing in the middle of context, especially should excessive whitespaces exist between words.  Something like `the        quick   brown fox` should appear to skip all spaces between tokens at once, not in multiple steps.

**Comment by DavidLRowe** [comment]
@darcywong00 Would you be able to review this PR? I don't think I'm qualified.

**Comment by jahorton** [comment]
For what it's worth, I've included a copy of the new wordbreaker code in https://github.com/keymanapp/keyman/pull/12229, which is then used in automated tests.  See https://github.com/keymanapp/keyman/blob/ec32e21bf4bd6f7d225126e34deaec7ff728e302/common/models/templates/test/test-tokenization.js#L529-L575.

**Comment by darcywong00** [comment]
> Recent changes in 18.0-alpha require an update to the custom wordbreaker I originally wrote for this model. 

So should these changes get in before we rebuild all the lexical-models in #246?
e.g. Should we bump the gcc-custom-breaker version (in the .kps) here or #246?

File: release/sil/sil.km.ggc/source/sil.km.ggc.model.ts (hunk 3 above)
Comment (darcywong00) [reply release/sil/sil.km.ggc/source/sil.km.ggc.model.ts:27]: Does the custom wordbreaker for sil.km.cnd.model.ts need a similar fix?

https://github.com/keymanapp/lexical-models/blob/a90fec5d323bc30f6fe1783ec604aab95f3d93a2/release/sil/sil.km.cnd/source/sil.km.cnd.model.ts#L13-L25

**Comment by jahorton** [comment]
> > Recent changes in 18.0-alpha require an update to the custom wordbreaker I originally wrote for this model.
> 
> So should these changes get in before we rebuild all the lexical-models in #246? e.g. Should we bump the gcc-custom-breaker version (in the .kps) here or #246?

Either order would likely be fine, but yeah, we'd probably need to bump the version again if done after #246, assuming it includes a version bump.  So... it wouldn't hurt to push this in before the "rebuild all".

**Review by mcdurdin — APPROVED** [review]

**Pull request merged**